R/SoilProfileCollection-operators.R

```diff
@@ -85,14 +85,14 @@ setMethod("[", signature(x = "SoilProfileCollection",
             }
 
             # extract all site and horizon data
-            h <- .as.data.frame.aqp(x@horizons, aqp_df_class(x))
+            h <- x@horizons
             s.all <- x@site
 
             # extract requested profile IDs
             p.ids <- s.all[[idname(x)]][unique(i)]
 
             # keep only the requested horizon data (filtered by profile ID)
-            h <- h[h[[idname(x)]] %in% p.ids,]
+            h <- .as.data.frame.aqp(h, aqp_df_class(x))[h[[idname(x)]] %in% p.ids,]
```

Comment:
Is this necessary every time the contents of `@horizons` is extracted / filtered based on row or column expressions?  I want to make sure that I understand the full implications of supporting 3 possible classes.

Reply:
I think that any time you do anything that could possibly be misinterpreted by data.table you need to call the method that will ensure that the data.table code is triggered on a data.table object. 

It is possible, though unlikely, that the metadata could say data.table but the user or another operation converted to data.table. Wrapping access to horizons in a call when you get it out and when you put it back in is the best way to ensure that.

Reply:
This back and forth I did here is weird. I think I was having an unrelated error and tried fiddling with this and did not change it back. Thanks for drawing my attention to it.  I can clean it up and call it when the slots (site and horizon) are accessed. This is all to support data.table. tbl_df is essentially a drop in replacement for data.frame.

Reply:
I'll have to think about this some more, during daylight hours. That is an incredibly verbose chunk of code required to perform some very basic operation on a `data.frame`-like object.  I'm starting to wonder if `data.table` would be more useful for on-the-fly operations vs. drop-in replacement. 

An example, would the internals of e.g. `horizonColorIndices` have to be re-written using `as.data.frame.aqp` and / or `.data.frame.j`?

Reply:
In that example, `horizons<-` handles it when merge `horizonColorIndices` back into the SPC

`horizonColorIndices` is written in a way that is unambiguous. `.data.frame.j` is only required for multi column subsetting in the j position. .as.data.frame.aqp is not needed either, because it is not a method that is internal to the object. Only the objects methods care about maintaining the data.frame object state. Random other methods do not care. They may take advantage of whatever class is there, or even convert to one, but I think that there is not as much "overhead" here as you are thinking. The major barrier is my obtuse code, which I will ponder how I can make clearer and document for you.

Reply:
Crap I deleted my comment. Here is the issue. 

horizonColorIndices will actually break if the object in the SPC is a data.table. It is unambiguous up till the last line.

```
  return(hz[,c(idname(p), hzidname(p), names(hz)[!names(hz) %in% names(horizons(p))])])
```
needs to be
```
  return(.data.frame.j(hz. c(idname(p), hzidname(p), names(hz)[!names(hz) %in% names(horizons(p))])))
```

Reply:
However, any potential conversion would happen internal to the object in e.g. `horizons<-` when the color indices were reinserted. The horizon slot should already be the same class, so only the class is checked internally by `.as.data.frame.aqp`

If you think we can switch over to data.table entirely I am willing to see what of these non-central methods will break -- maybe try changing the SPC object definition in a branch? I do not run the full panel of aqp tests on data.tables, so surely there are many more things that will break in the less exercised parts of the codebase

Reply:
The above is now fixed for horizonColorIndices.